### ISSUE TITLE
Common.mk wrong value on parameter/flag

### ISSUE THREAD

**Issue by cmoralesmx** [FLAMEGPU/FLAMEGPU#181]
I think there is a mistake in the value of the `code` parameter  in  
https://github.com/FLAMEGPU/FLAMEGPU/blob/801c54858e3f6ff2ac717412821de69d86fb644f/tools/common.mk#L263
It should be 
`code=sm_$(HIGHEST_SM) `  

Otherwise, I can't compile on Linux.

**Comment by ptheywood** [comment]
To enable JIT compilation of future architectures we pass the highest virtual architecture as arguments of both the virtual architecture (`arch=`) and real architecture (`code`). 

I.e to build code for GPUs of SM 20, 35 and above the gencode arguments would be: 
```
-gencode=arch=compute_20,code=sm_20
-gencode=arch=compute_35,code=sm_35
-gencode=arch=compute_35,code=compute_35
```

This is detailed in [section 5.6.1 of the current CUDA docs](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#just-in-time-compilation).

Which version of CUDA are you using to build FLAME GPU?

**Comment by cmoralesmx** [comment]
I just realized I may have miss identified the problem.  
First of all, I'm using a clean install of Ubuntu 16.04 with CUDA toolkit 8.0.  
After checking out the branch 'release-1.5', I attempted making the examples by doing
```sh
~/FLAMEGPU/examples/PedestrianNavigation$ make console SM=50 61
src/model/XMLModelFile.xml validates
nvcc -I ../../include -I src/model -I src/dynamic -I src/visualisation  -m64 -lineinfo -std=c++11  -Xcompiler -Wall -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o build/linux-x64/Release/main_console.cu.o -c src/dynamic/main.cu
LD_RUN_PATH='LD_RUN_PATH=$ORIGIN/../../../lib/x86_64-linux-gnu' nvcc   -m64 -lineinfo -std=c++11  -Xcompiler -Wall  -L../../lib/x86_64-linux-gnu  -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o ../../bin/linux-x64/Release_Console/PedestrianNavigation build/linux-x64/Release/io.cu.o build/linux-x64/Release/simulation.cu.o build/linux-x64/Release/main_console.cu.o
make: *** No rule to make target '61'. Stop
```
or 
```sh
~/FLAMEGPU/examples/PedestrianNavigation$ make console SM=50 60
src/model/XMLModelFile.xml validates
nvcc -I ../../include -I src/model -I src/dynamic -I src/visualisation  -m64 -lineinfo -std=c++11  -Xcompiler -Wall -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o build/linux-x64/Release/io.cu.o -c src/dynamic/io.cu
nvcc -I ../../include -I src/model -I src/dynamic -I src/visualisation  -m64 -lineinfo -std=c++11  -Xcompiler -Wall -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o build/linux-x64/Release/simulation.cu.o -c src/dynamic/simulation.cu
nvcc -I ../../include -I src/model -I src/dynamic -I src/visualisation  -m64 -lineinfo -std=c++11  -Xcompiler -Wall -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o build/linux-x64/Release/main_console.cu.o -c src/dynamic/main.cu
LD_RUN_PATH='LD_RUN_PATH=$ORIGIN/../../../lib/x86_64-linux-gnu' nvcc   -m64 -lineinfo -std=c++11  -Xcompiler -Wall  -L../../lib/x86_64-linux-gnu  -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o ../../bin/linux-x64/Release_Console/PedestrianNavigation build/linux-x64/Release/io.cu.o build/linux-x64/Release/simulation.cu.o build/linux-x64/Release/main_console.cu.o
make: *** No rule to make target '60'. Stop.
```
So, I did what I mentioned on my first message and the compilation succeeded. However, while trying to replicate the error I identified a fundamental difference in what I was doing.

Using the original code again, compilation succeeds if I just specify SM=60 or SM=61 (or any other single architecture) but fails whenever I try to specify 2 architectures in the way the help of the Makefile suggests  
https://github.com/FLAMEGPU/FLAMEGPU/blob/801c54858e3f6ff2ac717412821de69d86fb644f/examples/Makefile#L47

```sh
~/FLAMEGPU/examples/PedestrianNavigation$ make console SM=30 50
src/model/XMLModelFile.xml validates
nvcc -I ../../include -I src/model -I src/dynamic -I src/visualisation  -m64 -lineinfo -std=c++11  -Xcompiler -Wall -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o build/linux-x64/Release/io.cu.o -c src/dynamic/io.cu
nvcc -I ../../include -I src/model -I src/dynamic -I src/visualisation  -m64 -lineinfo -std=c++11  -Xcompiler -Wall -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o build/linux-x64/Release/simulation.cu.o -c src/dynamic/simulation.cu
nvcc -I ../../include -I src/model -I src/dynamic -I src/visualisation  -m64 -lineinfo -std=c++11  -Xcompiler -Wall -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o build/linux-x64/Release/main_console.cu.o -c src/dynamic/main.cu
LD_RUN_PATH='LD_RUN_PATH=$ORIGIN/../../../lib/x86_64-linux-gnu' nvcc   -m64 -lineinfo -std=c++11  -Xcompiler -Wall  -L../../lib/x86_64-linux-gnu  -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o ../../bin/linux-x64/Release_Console/PedestrianNavigation build/linux-x64/Release/io.cu.o build/linux-x64/Release/simulation.cu.o build/linux-x64/Release/main_console.cu.o
make: *** No rule to make target '50'. Stop.
```
However, it succeeds if I specify a single architecture or use a comma as the separator 
```sh
~/FLAMEGPU/examples/PedestrianNavigation$ make console SM=30,50,60,61
src/model/XMLModelFile.xml validates
nvcc -I ../../include -I src/model -I src/dynamic -I src/visualisation  -m64 -lineinfo -std=c++11  -Xcompiler -Wall -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o build/linux-x64/Release/io.cu.o -c src/dynamic/io.cu
nvcc -I ../../include -I src/model -I src/dynamic -I src/visualisation  -m64 -lineinfo -std=c++11  -Xcompiler -Wall -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o build/linux-x64/Release/simulation.cu.o -c src/dynamic/simulation.cu
nvcc -I ../../include -I src/model -I src/dynamic -I src/visualisation  -m64 -lineinfo -std=c++11  -Xcompiler -Wall -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o build/linux-x64/Release/main_console.cu.o -c src/dynamic/main.cu
LD_RUN_PATH='LD_RUN_PATH=$ORIGIN/../../../lib/x86_64-linux-gnu' nvcc   -m64 -lineinfo -std=c++11  -Xcompiler -Wall  -L../../lib/x86_64-linux-gnu  -gencode arch=compute_30,code=sm_30 -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_61,code=compute_61 -o ../../bin/linux-x64/Release_Console/PedestrianNavigation build/linux-x64/Release/io.cu.o build/linux-x64/Release/simulation.cu.o build/linux-x64/Release/main_console.cu.o
```
So the problem is not on the code but in the way the parameter should be specified?

By the way, in case this is relevant, after each attempt I am doing
```
~/FLAMEGPU/examples/PedestrianNavigation$ make clobber
```

**Comment by ptheywood** [comment]
#174 now includes several of the proposed improvements, which should be merged into release-1.5 shortly:

+ Documentation (`make help`) has been updated to encourage quote-contained space-separated lists
+ `SMS` can now be a comma separated value, or a space separated values
+ If `SM` is specified but `SMS` is not specified, the value of `SM` will be used and a warning raised
    + If `SM` and `SMS` are specified, the value of `SMS` will be used.
